src/lib/dnssd/tests/TestIncrementalResolve.cpp

```diff
@@ -0,0 +1,427 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <lib/dnssd/IncrementalResolve.h>
+
+#include <string.h>
+
+#include <lib/dnssd/minimal_mdns/records/IP.h>
+#include <lib/dnssd/minimal_mdns/records/Ptr.h>
+#include <lib/dnssd/minimal_mdns/records/ResourceRecord.h>
+#include <lib/dnssd/minimal_mdns/records/Srv.h>
+#include <lib/dnssd/minimal_mdns/records/Txt.h>
+#include <lib/support/UnitTestRegistration.h>
+
+#include <nlunit-test.h>
+
+using namespace chip;
+using namespace chip::Dnssd;
+using namespace mdns::Minimal;
+
+namespace {
+
+// Operational names must be <compressed-fabric>-<node>._matter._tcp.local
+constexpr uint8_t kTestOperationalName[] = "\0411234567898765432-ABCDEFEDCBAABCDE\07_matter\04_tcp\05local\00";
+
+// Commissionable names must be <instance>._matterc._udp.local
+constexpr uint8_t kTestCommissionableNode[] = "\020C5038835313B8B98\10_matterc\04_udp\05local\00";
+
+// Commissioner names must be <instance>._matterd._udp.local
+constexpr uint8_t kTestCommissionerNode[] = "\020C5038835313B8B98\10_matterd\04_udp\05local\00";
+
+// Server name that is preloaded by the `PreloadSrvRecord`
+constexpr uint8_t kTestServerName[] = "\04abcd\05local\00";
+
+void PreloadSrvRecord(nlTestSuite * inSuite, SrvRecord & record)
+{
+    // NOTE: record pointers persist beyond  this function, so
+    // this data MUST be static
+    static const uint8_t data[] = {
+        0,    12,                       // Priority
+        0,    3,                        // weight
+        0x12, 0x34,                     // port
+        4,    'a',  'b', 'c', 'd',      // QNAME part: abcd
+        5,    'l',  'o', 'c', 'a', 'l', // QNAME part: local
+        0,                              // QNAME ends
+    };
+
+    BytesRange packet(data, data + sizeof(data));
+    BytesRange srv_data(data, data + sizeof(data));
+
+    NL_TEST_ASSERT(inSuite, record.Parse(srv_data, packet));
+}
+
+/// Convenience method to have a  serialized QName:
+///
+/// static const uint8_t kData[] = "datahere\00";
+///  AsSerializedQName(kData);
+///
+/// NOTE: this MUST be using the string "" format to add an extra NULL
+/// terminator that this method discards.
```

Comment:
I realize this can have off-by-one but this is a test convenience helper. This is unit test code.

We do not have testing utilities for qnames. It would be easier if we would have a std::string equivalent for serializedqname and FullQName, but we do not. Adding them seemed tedious so I just took this function from other code dealing with qnames.

Reply:
Updated code mostly to not use this anymore. Testing code still needs this, but comments do not apply anymore so I removed them.